neo/sys/cpu.cpp

```diff
@@ -28,205 +28,414 @@ If you have questions concerning this license or the applicable additional terms
 #else // SDL1.2 or SDL2
-  #include <SDL_cpuinfo.h>
+#include <SDL_version.h>
+#include <SDL_cpuinfo.h>
+#endif
+
+// hack add missing instructions to SDL1.2
+#if SDL_VERSION_ATLEAST(2, 0, 0)
+#define SDL_SUPPORTS_ABOVE_SSE2
 #endif
 
 // MSVC header intrin.h uses strcmp and errors out when not set
 #define IDSTR_NO_REDIRECT
 
 #include "sys/platform.h"
 #include "framework/Common.h"
-
 #include "sys/sys_public.h"
 
-#ifdef NO_CPUID
-#undef NO_CPUID
-#endif
+// cpu registers
+#define _REG_EAX        0
+#define _REG_EBX        1
+#define _REG_ECX        2
+#define _REG_EDX        3
 
-#if defined(__GNUC__)
-	#if !defined(__i386__) && !defined(__x86_64__)
-		#define NO_CPUID
-	#endif
-#elif defined(_MSC_VER)
-	#if !defined(_M_IX86)
-		#define NO_CPUID
-	#endif
-#else
-#error unsupported compiler
+/*
+================
+CPUid
+================
+*/
+static inline void CPUid( int func, unsigned regs[4] ) {
+// guessing here but i reckon most unixes use the same function from cpuid.h
+#if defined(__unix__) || defined(MACOS_X) || defined(__APPLE__) || defined(__AROS__)
+    __cpuid( func, regs[_REG_EAX], regs[_REG_EBX], regs[_REG_ECX], regs[_REG_EDX] );
+#else /* msvc / mingw / clang */
+    int values[4] = { 0, 0, 0, 0 };
+
+    __cpuid( values, func );
+
+    regs[_REG_EAX] = static_cast<unsigned>( values[0] );
+    regs[_REG_EBX] = static_cast<unsigned>( values[1] );
+    regs[_REG_ECX] = static_cast<unsigned>( values[2] );
+    regs[_REG_EDX] = static_cast<unsigned>( values[3] );
 #endif
-
-#ifdef NO_CPUID
-void Sys_FPU_SetDAZ(bool enable) {
 }
 
-void Sys_FPU_SetFTZ(bool enable) {
-}
-#else
-
-#if defined(__GNUC__)
-static inline void CPUid(int index, int *a, int *b, int *c, int *d) {
-#if __x86_64__
-#	define REG_b "rbx"
-#	define REG_S "rsi"
-#elif __i386__
-#	define REG_b "ebx"
-#	define REG_S "esi"
-#endif
-	*a = *b = *c = *d = 0;
-
-	__asm__ volatile
-	(	"mov %%" REG_b ", %%" REG_S "\n\t"
-		"cpuid\n\t"
-		"xchg %%" REG_b ", %%" REG_S
-		:	"=a" (*a), "=S" (*b),
-			"=c" (*c), "=d" (*d)
-		: "0" (index));
-}
-#elif defined(_MSC_VER)
-#include <intrin.h>
-static inline void CPUid(int index, int *a, int *b, int *c, int *d) {
-	int info[4] = { };
-
-	// VS2005 and up
-	__cpuid(info, index);
-
-	*a = info[0];
-	*b = info[1];
-	*c = info[2];
-	*d = info[3];
+/*
+================
+IsAMD
+================
+*/
+static bool IsAMD( void ) {
+    char pstring[16];
+    char processorString[13];
+
+    // get name of processor
+    CPUid( 0, ( unsigned int * )pstring );
+    processorString[0] = pstring[4];
+    processorString[1] = pstring[5];
+    processorString[2] = pstring[6];
+    processorString[3] = pstring[7];
+    processorString[4] = pstring[12];
+    processorString[5] = pstring[13];
+    processorString[6] = pstring[14];
+    processorString[7] = pstring[15];
+    processorString[8] = pstring[8];
+    processorString[9] = pstring[9];
+    processorString[10] = pstring[10];
+    processorString[11] = pstring[11];
+    processorString[12] = 0;
```

Comment:
what's the point of copying pstring to processorString?

Reply:
not sure either it's from darkmod stgatilov might know ?

Reply:
null terminating pstring ? but there are better ways.

Reply:
sorry had to get some shuteye it was allmost 2.00 in the night here :)
not a young man anymore sadly xD

Reply:
for terminating some say 0 is sufficient if using C strings though i hear quite a lot prefer the special '\0' character.

tbh in a c++ project std::string would probably have been better.

Reply:
not sure how a std::string would help here, but for 0-termination `pstring[12] = 0;` should suffice?

Reply:
Looks like this code is directly from the original Doom3 GPL release: https://github.com/TTimo/doom3.gpl/blob/aaa855815ab484d5bd095f347163194ac569abcc/neo/sys/win32/win_cpu.cpp#L199-L223

still no idea why they did that, and I also think that it doesn't really matter if the users CPU is from AMD (original Doom3 used that to decide what clockrate a CPU needs to qualify for medium/high quality, but dhwem3 doesn't do that. https://github.com/TTimo/doom3.gpl/blob/aaa855815ab484d5bd095f347163194ac569abcc/neo/framework/Common.cpp#L2746-L2749)

Reply:
hmm could be right i admit i did not look at the original code release.
not sure myself why it was done like that myself.

yeah it probably makes less sense with the check today seing as ryzen supports anything intel does, not even sure if needed at the time of 3DNow since that was more like an extension of the MMX model.
probably could be used for determining unsupported cpu types but since there are little of those left and the ones left dont support playing doom3 anyway it might be prudent to remove it unless users actually want to look at cpu information in the doom3 console :)